### PR TITLE
enhancement(observability, file source): Improve file source observability

### PR DESCRIPTION
**Pull request by bruceg**
I think this is complete with the spec (complete with an update to the spec). Questions:

1. Does it really make sense to output the low-level I/O error as a tag on the metric? I made note in the spec to exclude that from the metrics, but can re-add it.
2. Does the `error_type` field make sense in the logs, given that they will include a message indicating the type of error?
3. Is there a good way of cross-referencing the cue docs in the component spec to avoid having to write out the list of error types multiple times?

Future sources should be _much_ less extensive than this.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

Related: #7973 Epic: #3900

### CODE REVIEW

**Comment by blt** [comment]
> Does it really make sense to output the low-level I/O error as a tag on the metric? I made note in the spec to exclude that from the metrics, but can re-add it.

I would probably end up finding this useful, personally.

**Review by blt — APPROVED** [review]
This seems fairly straightforward to me.

File: docs/specs/component.md [hunk 1]
```diff
@@ -208,11 +208,14 @@ implement since errors are specific to the component.
 
 * Properties
   * `error` - The specifics of the error condition, such as system error code, etc.
-  * `stage` - The stage at which the error occurred. MUST be one of
-    `receiving`, `processing`, or `sending`. This MAY be omitted from
-    being represented explicitly in the event data when the error may
-    only happen at one stage, but MUST be included in the emitted logs
-    and metrics as if it were present.
+  * `error_type` - The type of error condition. This MUST be one of the types
+    listed in the `error_type` enum list in the cue docs. TODO: List them here?
```
Comment (jszwedko) [inline docs/specs/component.md:212]: 🤔 I guess we could just link to the cue doc here. I'm not seeing another alternative at the moment aside from duplicating them.

**Review by jszwedko — APPROVED** [review]
Looks good! I'm on the fence about including the error message as a tag on the metric. I don't have a good sense for if that is common or not.

File: website/cue/reference/components/sources/internal_metrics.cue [hunk 2]
```diff
@@ -1073,22 +1089,32 @@ components: sources: internal_metrics: {
 			required:    true
 			examples: ["http://localhost:8080/server-status?auto"]
 		}
+		_error: {
+			description: "The specifics of the error condition, such as system error code, etc."
+			required:    true
+			examples: ["File not found", "Connection dropped by remote"]
+		}
```
Comment (jszwedko) [inline website/cue/reference/components/sources/internal_metrics.cue:1096]: Are there are concerns about downstream services handling tag values like this? After further thought, they do seem a little non-standard with what I'm familiar with. What would you think of leaving this tag out until someone asks for it?

Comment (bruceg) [reply website/cue/reference/components/sources/internal_metrics.cue:1096]: That was my concern as well, that and a potential for cardinality issues.

File: docs/specs/component.md (hunk 1 above)
Comment (jszwedko) [reply docs/specs/component.md:212]: Just a note to resolve this TODO

File: docs/specs/component.md [hunk 3]
```diff
@@ -208,11 +208,14 @@ implement since errors are specific to the component.
 
 * Properties
   * `error` - The specifics of the error condition, such as system error code, etc.
```
Comment (jszwedko) [inline docs/specs/component.md:210]: Should we drop this property too given https://github.com/vectordotdev/vector/pull/9205/commits/89f6974de69a4370c0505e0ec69d30d1e31548e8 ? Or perhaps other metrics would have it.

File: website/cue/reference/components/sources/internal_metrics.cue [hunk 4]
```diff
@@ -1073,22 +1089,32 @@ components: sources: internal_metrics: {
 			required:    true
 			examples: ["http://localhost:8080/server-status?auto"]
 		}
+		_error: {
+			description: "The specifics of the error condition, such as system error code, etc."
+			required:    true
+			examples: ["File not found", "Connection dropped by remote"]
```
Comment (jszwedko) [inline website/cue/reference/components/sources/internal_metrics.cue:1095]: Should we drop this tag? Or do you anticipate other metrics having it? Should we remove these examples at least?

**Review by jszwedko — COMMENTED** [review]
A few follow-up questions about removing the `error` tag, but otherwise this looks good.

File: website/cue/reference/components/sources/internal_metrics.cue (hunk 4 above)
Comment (bruceg) [reply website/cue/reference/components/sources/internal_metrics.cue:1095]: Yes, I think if we should drop it. With the `error` tag not used here, there are no uses, so no need to introduce it.

File: docs/specs/component.md (hunk 3 above)
Comment (bruceg) [reply docs/specs/component.md:210]: No, the property is still used by the log, just not the metric.

**Pull request merged**